### PR TITLE
Remove SPLIT_BINARY

### LINKED ISSUE #37847: Remove "CLICKHOUSE_SPLIT_BINARY" option.
People said it does not work.

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Remove the option to build ClickHouse tools as separate executable programs. This fixes #37847.

### CODE REVIEW

**Comment by pkit** [comment]
@alexey-milovidov you should probably remove `-DUSE_STATIC_LIBRARIES=0 -DSPLIT_SHARED_LIBRARIES=1` too
As for example `flatbuffers` does not have shared lib, thus compilation still breaks...

**Comment by azat** [comment]
>@alexey-milovidov you should probably remove -DUSE_STATIC_LIBRARIES=0 -DSPLIT_SHARED_LIBRARIES=1 too

This is actually different, shared libraries compiles faster this can be useful for development, although not for production.

>As for example flatbuffers does not have shared lib, thus compilation still breaks...

It does - https://github.com/ClickHouse/ClickHouse/blob/04944ff6f51903c52daac4c28960eaea16c8c0b4/contrib/arrow-cmake/CMakeLists.txt#L83

**Comment by pkit** [comment]
@azat nope, fails for me with "cannot find -lflatbuffers_shared"

File: docs/en/development/developer-instruction.md [hunk 1]
```diff
@@ -281,7 +281,7 @@ Most probably some of the builds will fail at first times. This is due to the fa
 
 ClickHouse is normally statically linked into a single static `clickhouse` binary with minimal dependencies. This is convenient for distribution, but it means that for every change the entire binary needs to be re-linked, which is slow and inconvenient for development. As an alternative, you can instead build dynamically linked shared libraries and separate binaries `clickhouse-server`, `clickhouse-client` etc., allowing for faster incremental builds. To use it, add the following flags to your `cmake` invocation:
 ```
--DUSE_STATIC_LIBRARIES=0 -DSPLIT_SHARED_LIBRARIES=1 -DCLICKHOUSE_SPLIT_BINARY=1
+-DUSE_STATIC_LIBRARIES=0 -DSPLIT_SHARED_LIBRARIES=1
```
Comment (rschu1ze) [inline docs/en/development/developer-instruction.md:284]: Part "... and separate binaries clickhouse-server, clickhouse-client etc." (l. 282) can now be removed.

File: docker/packager/packager [hunk 2]
```diff
@@ -261,8 +261,7 @@ def parse_env_variables(
 
     if split_binary:
         cmake_flags.append(
-            "-DUSE_STATIC_LIBRARIES=0 -DSPLIT_SHARED_LIBRARIES=1 "
-            "-DCLICKHOUSE_SPLIT_BINARY=1"
+            "-DUSE_STATIC_LIBRARIES=0 -DSPLIT_SHARED_LIBRARIES=1"
```
Comment (rschu1ze) [inline docker/packager/packager:264]: Variable "split_binary" can now be renamed to "shared_libraries".

**Comment by rschu1ze** [comment]
`SPLIT_BINARY` and `SHARED_LIBRARIES` are both developer options, thus "use at your own risk". Even if flatbuffer or another lib does not build with `SHARED_LIBRARIES`,  this option is still useful IMHO and the flatbuffers build can be easily disabled using `ENABLE_LIBRARIES`, see file "contrib/arrow-cmake/CMakeLists.txt"

**Comment by pkit** [comment]
@rschu1ze unfortunately my whole stuff heavily depends on arrow to work properly. But fortunately I use a good desktop that compiles CH from scratch in 25 min.

**Comment by azat** [comment]
>@azat nope, fails for me with "cannot find -lflatbuffers_shared"

@pkit there is something specific in your env, since it works on [CI](https://s3.amazonaws.com/clickhouse-builds/39520/071374b1525b9b7c6ae51a41e0cf3bd62b49e700/binary_splitted/build_log.log) (see for `flatbuffers_shared` and `libflatbuffers.so`)

**Comment by josabraham** [comment]
hi, building clickhouse with the following option throws the following error
create build directory and move into it they use following command

cmake .. -DUSE_STATIC_LIBRARIES=0 -DSPLIT_SHARED_LIBRARIES=1

and build using ninja command

ClickHouse/programs/su/clickhouse-su.cpp:1:10: fatal error: 'Common/Exception.h' file not found
#include <Common/Exception.h>

**Comment by rschu1ze** [comment]
@josabraham The problem sounds unrelated to SPLIT_BINARY. Did you build from this PR?

On my system, `cmake -S . -B build -DUSE_STATIC_LIBRARIES=0 -DSPLIT_SHARED_LIBRARIES=1` + `cmake --build build programs/su/CMakeFiles/clickhouse-su-lib.dir/clickhouse-su.cpp.o` (build target extracted from "build/build.ninja") works just fine.

Suggest to open a separate issue with a detailed description of the build steps and the error.

**Comment by josabraham** [comment]
> 

hi @rschu1ze it was my mistake . I was building using master branch.it worked with [no-split-binary](https://github.com/ClickHouse/ClickHouse/tree/no-split-binary) branch.Thank you.

**Comment by rschu1ze** [comment]
Strange effect in CI [Split Build Smoke Test](https://s3.amazonaws.com/clickhouse-test-reports/39520/d07970de7aa3c06572eac7f1bee866c97d75dee7/split_build_smoke_test.html) job: `/unpacked/clickhouse-server: error while loading shared libraries: libharmful.so: cannot open shared object file: No such file or directory`.

At least on the CMake / source code side, everything is fine. I could also not reproduce the failure locally with Debug or Release builds but a lot more magic happens in the central builds. The error is produced by file "docker/test/split_build_smoke_test/run.sh" which also sets "LD_LIBRARY_PATH" but for that it should not really matter if the binaries are split or not.

EDIT: It actually does seem to have to do with "libharmful.so". I ran `ldd clickhouse-server` on
- a local build of this PR equivalent to a "split build", i.e. `-DUSE_STATIC_LIBRARIES=0 -DSPLIT_SHARED_LIBRARIES=1` and I find "libharmful" as a shared object dependency
- a local build of master which is equivalent to the previous definition of a "split build", i.e. `-DUSE_STATIC_LIBRARIES=0 -DSPLIT_SHARED_LIBRARIES=1 -DSPLIT_BINARY=1` and I do *not* find "libharmful" as a shared object dependency.

Indeed, looking at the changes in "programs/CMakeLists.txt", `libharmful` was not been linked for `-DSPLIT_BINARY=1`. I don't know if that was by purpose or accidentally.

So, libharmful is linked, and the question is why it cannot be found at server startup. File "docker/packager/binary/build.sh" packs various binaries and libraries into an archive which is unpacked by "split_build_smoke_test/run.sh" later on. As can be seen [here](https://pipelines.actions.githubusercontent.com/serviceHosts/70b90959-6b3c-4d1f-b928-14fabc062a3b/_apis/pipelines/1/runs/111482/signedlogcontent/112?urlExpires=2022-07-30T21%3A37%3A49.9509625Z&urlSigningMethod=HMACV1&urlSignature=0y8LlJKPiAX1egfM2twtmy3yezOwQ0ESInd8sjbTX9I%3D), libharmful is actually packed:
```
[...]
2022-07-29T13:47:59.7952362Z Jul 29 13:47:59 + find . -name '*.so' -print -exec mv '{}' /output ';'
[...]
2022-07-29T13:48:02.5959994Z Jul 29 13:48:02 ./base/harmful/libharmful.so
```
So I basically have no idea what is going wrong here 😞 

One option would be to not link libharmful when `-DUSE_STATIC_LIBRARIES=0 -DSPLIT_SHARED_LIBRARIES=1` and hope for the best. It's hacky but since it's a special developer option that maybe okay.

**Comment by alexey-milovidov** [comment]
> One option would be to not link libharmful when -DUSE_STATIC_LIBRARIES=0 -DSPLIT_SHARED_LIBRARIES=1 and hope for the best. It's hacky but since it's a special developer option that maybe okay.

Yes, it will be ok.

**Comment by azat** [comment]
>Strange effect in CI [Split Build Smoke Test](https://s3.amazonaws.com/clickhouse-test-reports/39520/d07970de7aa3c06572eac7f1bee866c97d75dee7/split_build_smoke_test.html) job: /unpacked/clickhouse-server: error while loading shared libraries: libharmful.so: cannot open shared object file: No such file or directory.

This is not CI by itself, it is `checkHarmfulEnvironmentVariables` - https://github.com/ClickHouse/ClickHouse/blob/c9e685030625f749c7564fc861d8aa1aab8f5e60/programs/main.cpp#L349-L363

And since you removed split binary, then it is used always, and because `unpacked` is not included into `DT_RUNPATH`/`DT_RPATH` after re-exec it cannot find the library:

```
$ LD_LIBRARY_PATH=unpacked strace -k -s1000 -feexecve unpacked/clickhouse-server
execve("unpacked/clickhouse-server", ["unpacked/clickhouse-server"], 0x7fffffffe218 /* 44 vars */) = 0
 > /usr/lib/x86_64-linux-gnu/ld-2.33.so() [0x1050]
execve("unpacked/clickhouse-server", ["unpacked/clickhouse-server"], 0x7fffffffe228 /* 44 vars */) = 0
 > /usr/lib/x86_64-linux-gnu/libc-2.33.so(execve+0x7) [0xc6f77]
 > /usr/lib/x86_64-linux-gnu/libc-2.33.so(execlp+0x2c5) [0xc7695]
 > /work1/azat/ch/tmp/39520/unpacked/clickhouse(main+0x744) [0x20504]
unpacked/clickhouse-server: error while loading shared libraries: libharmful.so: cannot open shared object file: No such file or directory
+++ exited with 127 +++
```

Locally, I guess, you haven't tried it in the same way as it is done on CI, for example if you will unpack everything to `CWD` and run from it `clickhouse` it will work, because `.` is included into `DT_RPATH`

By the way, this means that now `programs/BINARY_NAME.cpp` can be removed?

**Comment by rschu1ze** [comment]
Great find, thanks! Will disable the check for split builds.

> By the way, this means that now programs/BINARY_NAME.cpp can be removed?

What exactly do you mean by that? The only ".cpp" file in "programs" is "main.cpp".

**Comment by azat** [comment]
>What exactly do you mean by that? The only ".cpp" file in "programs" is "main.cpp".

Ops, sorry, I mean't `programs/BINARY/clickhouse-BINARY.cpp`

**Comment by rschu1ze** [comment]
> Ops, sorry, I mean't `programs/BINARY/clickhouse-BINARY.cpp`

Each binary has it's own `main()` and depending on the binary, there is quite some custom code. File "programs/BINARY/clickhouse-BINARY.cpp" looks like:

```
int mainEntryClickHouseBINARY(int argc, char ** argv);                                
int main(int argc_, char ** argv_) { return mainEntryClickHouseBINARY(argc_, argv_); }
```

All such `main()` methods could theoretically be moved into "programs/main.cpp" (and ifdef'd) but that wouldn't make the code more beautiful, I guess.

**Comment by rschu1ze** [comment]
"SplitBuildSmokeTest" survived this PR --> merging.

**Pull request merged**